File: 2023/nim/p02.nim
import strutils, sequtils, re, tables, math

const colors = {"red": 0, "green": 1, "blue": 2}.toTable
let
    input = readFile("../data/02.txt").splitLines
    games = input.map(proc(line: auto): auto =
        let
            parts = line.split(re"[:;]")
            id = parts[0].split(' ')[1].parseInt
            res = parts[1..^1].map(proc(it: auto): auto =
                var rgb = @[0, 0, 0]
                for desc in it.strip.split(','):
                    let x = desc.strip.split(' ', 2)
                    rgb[colors[x[1]]] = x[0].parseInt
                rgb
            )
        (id, res)
    )

# part 1
echo games.map(proc(game: auto): auto =
    let valid = game[1].allIt(it[0] <= 12 and it[1] <= 13 and it[2] <= 14)
    if valid: game[0] else: 0
    )
    .sum

# part 2
echo games.map(proc(game: auto): auto =
    let
        r = game[1].mapIt(it[0]).max
        g = game[1].mapIt(it[1]).max
        b = game[1].mapIt(it[2]).max
    r*g*b
    )
    .sum

File: 2023/ruby/p02.rb
input = IO.foreach('../data/02.txt').map(&:strip)

games = input.map{|line| 
  parts = line.split(/[:;]/)
  id = parts[0].split(' ')[1].to_i
  res = parts[1..-1].map{ 
    comps = _1.strip.split(/[, ]+/)
    h = Hash[*comps.reverse]
    [h["red"].to_i, h["green"].to_i, h["blue"].to_i]
  }
  [id, res]
}

# part 1
p games.map{|id, parts|
  valid = parts.all?{|r,g,b| r <= 12 && g <= 13 && b <= 14}
  valid ? id : 0
}.sum

# part 2
p games.map{|id, parts|
  (0..2).map{|i| parts.map{_1[i]}.max}.reduce(:*)
}.sum
